### PR TITLE
[mysql] Remove the finished snapshot splits for the binlog split when restoring from the checkpoint

### LINKED ISSUE #2282: [Bug] 修改tableList然后从保存点启动，增量无法同步
### Search before asking

- [X] I searched in the [issues](https://github.com/ververica/flink-cdc-connectors/issues) and found nothing similar.

### Flink version

1.17.1

### Flink CDC version

2.40

### Database and its version

mysql 5.7

### Minimal reproduce step

step1: 设置table list为["egenie_goods.goods","egenie_goods.test","egenie_goods.goods_business_props"]
step2: 设置保存点，使用yarn app --kill杀掉job
step3: 设置table list为["egenie_goods.goods","egenie_goods.test","egenie_goods.goods_business_props_\\d+$"], 从第二步保存点启动，发现全量完成后，增量数据无法同步。egenie_goods.goods，egenie_goods.test和egenie_goods.goods_business_props_0,等所有的增量都不行。

随便说下, 如果修改source的并行度为1，一切又正常。
DataStreamSource<ChangeRecord> map = env.fromSource(mySqlSource, WatermarkStrategy.noWatermarks(), "mysqlSource").setParallelism(1);

### What did you expect to see?

修改tableList，全量结束后，能正常增量同步。

### What did you see instead?

增量同步不成功， 日志一直在刷：
2023-07-10 14:57:13,409 INFO  com.ververica.cdc.connectors.mysql.source.reader.MySqlSourceReader [] - Source reader 0 fills metadata of group 145 to binlog split

### Anything else?

_No response_

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by ruanhang1993**
Fix #2282 .

### CODE REVIEW

File: flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/reader/MySqlSourceReader.java [hunk 1]
```diff
@@ -203,6 +204,10 @@ private boolean isNewlyAddedTableSplitAndBinlogSplit(
 
     @Override
     public void addSplits(List<MySqlSplit> splits) {
+        addSplits(splits, false);
+    }
+
+    private void addSplits(List<MySqlSplit> splits, boolean selfUpdate) {
```
Comment (leonardBang) [inline flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/reader/MySqlSourceReader.java:210]: it's hard to understand `selfUpdate`. How about 
```suggestion
        /**
     * Adds a list of splits for this reader to read.
     *
     * @param splits the splits to add.
     * @param checkTablesForBinlogSplit to check the captured table list change or not, it should be
     *     true for reader which is during restoration from a checkpoint or savepoint.
     */
    private void addSplits(List<MySqlSplit> splits, boolean checkTableChangeForBinlogSplit) {
    ...
       if (checkTableChangeForBinlogSplit) { 
       filterOutdatedSplitInfos(...)
    
```

File: flink-connector-mysql-cdc/src/test/java/com/ververica/cdc/connectors/mysql/source/NewlyAddedTableITCase.java [hunk 2]
```diff
@@ -310,6 +331,277 @@ public void testRemoveTable() throws Exception {
...
+        CollectStreamSink<RowData> sink = new CollectStreamSink<>(stream, factory);
+        sink.name("Data stream collect sink");
+        stream.getExecutionEnvironment().addOperator(sink.getTransformation());
+        CollectResultIterator<RowData> iterator =
+                new CollectResultIterator(
+                        operator.getOperatorIdFuture(),
+                        serializer,
+                        accumulatorName,
+                        stream.getExecutionEnvironment().getCheckpointConfig());
+        //        iterator.setJobClient(jobClient);
```
Comment (leonardBang) [inline flink-connector-mysql-cdc/src/test/java/com/ververica/cdc/connectors/mysql/source/NewlyAddedTableITCase.java:567]: useless?

File: flink-connector-mysql-cdc/src/test/java/com/ververica/cdc/connectors/mysql/source/NewlyAddedTableITCase.java [hunk 3]
```diff
@@ -310,6 +331,277 @@ public void testRemoveTable() throws Exception {
...
+                                    tableInThisRound),
+                            String.format(
+                                    "+I[121, user_8, Shanghai, 123567891234, %s]",
+                                    tableInThisRound),
+                            String.format(
+                                    "+I[1017, user_18, Shanghai, 123567891234, %s]",
+                                    tableInThisRound),
+                            String.format(
+                                    "+I[1018, user_19, Shanghai, 123567891234, %s]",
+                                    tableInThisRound));
```
Comment (leonardBang) [inline flink-connector-mysql-cdc/src/test/java/com/ververica/cdc/connectors/mysql/source/NewlyAddedTableITCase.java:498]: If we rename `tableInThisRound` to `changedTable` and use static import from `String.format` to `format`, the code will be more readable.

**Review by leonardBang — COMMENTED** [review]
Thanks @ruanhang1993  for the fix, I left some comments.

**Review by leonardBang — APPROVED** [review]
Thanks @ruanhang1993  for the update, LGTM

File: flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/reader/MySqlSourceReader.java [hunk 4]
```diff
@@ -224,6 +236,20 @@ public void addSplits(List<MySqlSplit> splits) {
                 }
             } else {
                 MySqlBinlogSplit binlogSplit = split.asBinlogSplit();
+                // When restore from a checkpoint, the finished split infos may contain some splits
+                // for the deleted tables.
+                // We need to remove these splits for the deleted tables at the finished split
+                // infos.
+                if (!checkTableChangeForBinlogSplit) {
```
Comment (leonardBang) [inline flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/reader/MySqlSourceReader.java:243]: wait, here should be `if(checkTableChangeForBinlogSplit)`, we should update the code where this method is called

**Review by leonardBang — APPROVED** [review]

File: flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/reader/MySqlSourceReader.java (hunk 4 above)
Comment (ruanhang1993) [reply flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/reader/MySqlSourceReader.java]: Yes, I have fixed this part in the newest commit.

**Pull request merged**